apache/trafficcontrol
# Idempotentiate create_tables.sql

## What does this PR (Pull Request) do?


- [x] This PR closes #4984
- This PR makes it so that, even if `./admin load_schema` is run more than once (for example, installing TO, with a shared database, on more than 1 host simultaneously), an SQL error should not result.

## Which Traffic Control components are affected by this PR?

- Traffic Ops (initial `create_tables.sql` seed data)

## What is the best way to verify this PR?

1. Run the Traffic Ops DB tests
    ```shell script
    ./pkg traffic_ops_build;
    cd traffic_ops_db/test/docker;
    cp ../../../dist/traffic_ops-5.0.0-10961.9250b20e.el7.x86_64.rpm traffic_ops.rpm;
    docker-compose up --build;
    ```

2. Check that the resultant schema is identical to the one we would have gotten from `create_tables.sql` previously.
    ```shell script
    import_and_dump_schema() {
        output_file="$1";
        <<'DOCKER_COMMANDS' docker run --rm -i \
            -v "$(pwd)/traffic_ops/app/db/create_tables.sql:/create_tables.sql" \
            -e POSTGRES_USER=traffic_ops \
            -e POSTGRES_PASSWORD=twelve \
            -e PGPASSWORD=twelve \
            -e POSTGRES_DB=traffic_ops \
            postgres:9.6-alpine bash >"$output_file";
        trap 'echo "Error on line ${LINENO} of ${0}" >/dev/stderr; exit 1' ERR;
        set -o errexit;
        /docker-entrypoint.sh postgres >/dev/stderr &
        echo 'Waiting for Postgres to start...' >/dev/stderr;
        sleep 10;
        psql -Utraffic_ops -fcreate_tables.sql >/dev/stderr;
        pg_dump -Utraffic_ops;
    DOCKER_COMMANDS
    };
    git checkout origin/master;
    import_and_dump_schema upstream.sql;
    git checkout zrhoffman/postinstall-twice;
    import_and_dump_schema pr.sql;
    wc -l upstream.sql pr.sql; # Should be 4429 lines
    openssl sha256 upstream.sql pr.sql; # Hash should be 794fd7a23fa690eec67f6cd1257a9a8d900ac59edbdc487119ef73023a938d84
    ```

## The following criteria are ALL met by this PR


- [x] You should never actually run `create_tables.sql` twice, no documentation necessary
- [x] This PR ensures that database migration sequence is correct
